README
-----------------------------------------------------

     =========================================================
     Geant4 - an Object-Oriented Toolkit for Simulation in HEP
     =========================================================

                Simple Absorbed Silicon Detector
                            SASiDet
                           based on
                            TestEm5
                            -------
 
 How to study the transmission, absorption and reflection of particles through
 a single, thin or thick, layer of material.
 In particular, the effects of the multiple scattering can be plotted.

 Macro files used for Milan Roberson's Senior Thesis:
 All simulate 1,000,000 x-rays, using a flat input spectrum from 2-25 keV
 thesis_abs_away_const.mac
	Absorber is present, Aluminized side is facing away from detector
 thesis_abs_towards_const.mac
 	Absorber is present, Aluminized side is facing towards detector
 thesis_no_abs_const.mac
	Absorber is not present.

 1- GEOMETRY DEFINITION

 The "World" is a 30cm side cube made of air.
 The absorber (if present) is 1cm in the positive X direction from the center of the World,
 and is composed of 1000A of Al, 127 microns of Kapton, 100A of ITO. Its orientation (which
 side is facing the detector) can be changed by command prior to Run Initialization.
 The detector is modeled after the AmpTek XR-100CR Si detector, and is placed 10cm in the
 positive X direction from the center of the World.

 This default geometry can be changed in DetectorConstruction

 2- PHYSICS LIST

 Physics lists are based on modular design. Several modules are instantiated:
 1. Transportation
 2. EM physics
 3. Decays
 4. StepMax - for step limitation

 EM physics builders can be local (eg. in this example) or from G4 kernel
 physics_lists subdirectory.

 Local physics builders:	 
 - "local"	 standard EM physics with current 'best' options setting
                 these options are explicited in PhysListEmStandard
 - "standardSSM" standard EM physics with alternative single Coulomb 
                 scattering model instead of multiple scattering.

 From geant4/source/physics_lists/builders:	 
 - "emstandard_opt0" recommended standard EM physics for LHC
 - "emstandard_opt1" best CPU performance standard physics for LHC
 - "emstandard_opt2" similar fast simulation
 - "emstandard_opt3" best standard EM options - analog to "local" above
 - "emstandard_opt4" best current advanced EM options standard + lowenergy
 - "emstandardWVI" standard EM physics and WentzelVI multiple scattering
 - "emstandardSS"  standard EM physics and single scattering model
 - "emlivermore"  low-energy EM physics using Livermore data
 - "empenelope"   low-energy EM physics implementing Penelope models
 - "emlowenergy"  low-energy EM physics implementing experimental 
                  low-energy models

 Physics lists and options can be (re)set with UI commands

 Please, notice that options set through G4EmProcessOptions are global, eg
 for all particle types. In G4 builders, it is shown how to set options per
 particle type.

 3- AN EVENT : THE PRIMARY GENERATOR

 The primary kinematic consists of a single particle which hits the absorber 
 perpendicular to the input face. The type of the particle and its energy are 
 set in the PrimaryGeneratorAction class, and can be changed via the G4 
 build-in commands of G4GeneralParticleSource class (see the macros provided with this 
 example).

 In addition one can choose randomly the impact point of the incident particle.
 The interactive command is built in PrimaryGeneratorMessenger class.

 4- VISUALIZATION

 The Visualization Manager is set in the main().
 The initialisation of the drawing is done via the commands in vis.mac
 In interactive session:
 PreInit or Idle > /control/execute vis.mac

 The example has a default view which is a longitudinal view of the detector.

 The tracks are drawn at the end of event, and erased at the end of run.
 Optionally one can choose to draw all particles, only the charged, or none.
 This command is defined in EventActionMessenger class.

 5- TRACKING

 During the tracking, one can keep or not the secondaries : see StackingAction
 class and its Messenger (StackingMessenger).
 One can also limit 'by hand' the step lenght of the particle. As an example,
 this limitation is implemented as a 'full' process : see StepMax class and its
 Messenger. The 'StepMax process' is registered in the Physics List. 

 6- DETECTOR RESPONSE
 
 At the end of a run, from the histogram(s), one can study different
 physics quantities such as :
 - energy deposit in the absorber,
 - energy spectrum of secondaries at creation, 
 - energy spectrum and angle distribution of particles at exit,
 - transmission and backscattering coefficients,
 -  ...

 7- List of the built-in histograms
 ----------------------------------

 The test contains more than 60 built-in 1D histograms, which are managed by
 G4AnalysisManager class and its Messenger. The histos can be individually activated
 with the command :
 /analysis/h1/set id nbBins  valMin valMax unit 
 where unit is the desired unit for the histo (MeV or keV, deg or mrad, etc..)
 (see the macros xxxx.mac).

	1	"energy deposit in absorber"
	2	"energy of charged secondaries at creation"
	3	"energy of neutral secondaries at creation"	
	4	"energy of charged at creation (log10(Ekin))"
	5	"energy of neutral at creation (log10(Ekin))"	
	6   "x_vertex of charged secondaries (all)"
	7   "x_vertex of charged secondaries (not absorbed)"
	10	"(transmit, charged) : kinetic energy at exit of world"
	11	"(transmit, charged) : ener fluence: dE(MeV)/dOmega"	
	12	"(transmit, charged) : space angle dN/dOmega"
	13	"(transmit, charged) : projected angle at exit of world"
	14	"(transmit, charged) : projected position at exit of world"
	15	"(transmit, charged) : radius at exit of world"	
	20	"(transmit, neutral) : kinetic energy at exit of world"
	21	"(transmit, neutral) : ener fluence: dE(MeV)/dOmega"	
	22	"(transmit, neutral) : space angle dN/dOmega"
	23	"(transmit, neutral) : projected angle at exit of world"
	30	"(reflect , charged) : kinetic energy at exit of world"
	31	"(reflect , charged) : ener fluence: dE(MeV)/dOmega"	
	32	"(reflect , charged) : space angle dN/dOmega"
	33	"(reflect , charged) : projected angle at exit of world"
	40	"(reflect , neutral) : kinetic energy at exit of world"
	41	"(reflect , neutral) : ener fluence: dE(MeV)/dOmega"	
	42	"(reflect , neutral) : space angle dN/dOmega"
	43	"(reflect , neutral) : projected angle at exit of world"
    50  "energy of Auger e- at creation"
    51  "energy of fluorescence gamma at creation"
    52  "energy of Auger e- at creation (log scale)"
    53  "energy of fluorescence gamma at creation (log scale)"
    54  "energy of PIXE Auger e- at creation"
    55  "energy of PIXE gamma at creation"
    56  "energy of PIXE Auger e- at creation (log scale)"
    57  "energy of PIXE gamma at creation (log scale)"
    58  "energy of G4DNA Auger e- at creation"
    59  "energy of G4DNA gamma at creation"
    60  "energy of G4DNA Auger e- at creation (log scale)"
    61  "energy of G4DNA gamma at creation (log scale)"

 The histograms can be viewed using ROOT or PAW.

 One can control the name of the histograms file with the command:
 /analysis/setFileName  name  (default sasidet)

 It is possible to choose the format of the histogram file : root (default),
 hbook, xml, csv, by using namespace in HistoManager.hh

 It is also possible to print selected histograms on an ascii file:
 /analysis/h1/setAscii id
 All selected histos will be written on a file name.ascii  (default sasidet) 

 8- GEANT4/GEANT3/DATA COMPARISON

 A Geant4/Geant3/exp. data comparison is given here for a few cases.
 These cases can be classified as follow:
 - e-/e+ incident particles versus protons and others. 
 - 3 energy regimes: low: < 1MeV; medium: 1MeV -> few 10MeV; high: > 100MeV
 
 We indicate here the corresponding macros.
 
 	      |	low energy   |	medium energy	|  high energy
	--------------------------------------------------------
	      | acosta.mac   |                  |
	e-+   |	berger.mac   |	hanson.mac      |
	      |	hunger.mac   |	kulchi.mac      |
	      | tavola.mac   |			|
	--------------------------------------------------------
	others|	bichsel.mac  | 	vincour.mac	|  shen1.mac shen2.mac
	      | 	     |	gottsch.mac	|  tramu.mac
	--------------------------------------------------------

 The relevant part of the GEANT3 code is in the subdirectory geant3 together 
 with the xxxx.dat input files.
	
 9- HOW TO START ?
 
 - execute SASiDet in 'batch' mode from macro files e.g.
 	% $(G4INSTALL)/bin/$(G4SYSTEM)/SASiDet   myMacro.mac
		
 - execute SASiDet in 'interactive' mode with visualization e.g.
 	% $(G4INSTALL)/bin/$(G4SYSTEM)/SASiDet
	Then type your commands, for instance :
	Idle> control/execute vis.mac
	Idle> run/beamOn 5
	....


History
----------------------------------------------------

     =========================================================
     Geant4 - an Object-Oriented Toolkit for Simulation in HEP
     =========================================================

                    SASiDet History file
                    ------------------------
This file should be used by the G4 example coordinator to briefly
summarize all major modifications introduced in the code and keep
track of all tags.

     ----------------------------------------------------------
     * Reverse chronological order (last date on top), please *
     ----------------------------------------------------------

20-07-20 Milan Roberson

- Renamed to Simple Absorbed Silicon Detector (SASiDet) from TestEm5
- Adapted to simulate absorption through Al-Kapton-ITO film and detection by Si
  detector
- Removed unused code.

15-01-19 V.Ivant (testem5-V10-05-00)
- StackinAction - fixed all log histograms, which where filled in a wrong 
                  way
- PhysListEm19DStandard - use 3-gamma annihilation model

09-09-18 S.Incerti (testem5-V10-04-07)
- updated dna.mac

18-05-18 V.Ivant (testem5-V10-04-06)
18-05-18 V.Ivant (testem5-V10-04-05)
- added test of the new ion ionisation model, which is included in 
  the new physics constructor PhysListEm19DStandard, atima.mac 
  macro file is also added to compare with known data

09-05-18 B.Morgan (testem5-V10-04-04)
- TestEm5.cc : include G4Types before use of G4MULTITHREADED. For forward
  compatibility with move to #defines over -D for G4 preprocessor
  symbols.

08-04-18 mma (testem5-V10-04-03)
- TestEm5.cc : purely cosmetic

20-03-18 V.Ivant (testem5-V10-04-02)
- StepMax - use thread safe instantiation, mesenger merged with 
    PhysicsListMessenger
- DetectorConstruction - improved method names
- PhysicsList - make Opt4 default, removed local instantiation of G4Decay
- Use MixMax default random number generator
- removed G4VIS_USE and G4UI_USE

08-03-18 V.Ivant (testem5-V10-04-01)
- PhysListEm5DStandard - experimental PhysList with the new 
    5D gamma conversion model
- PhysListEm5DStandard - clean-up for 10.4
- PhysListEmStandardSSM - removed as obsolete

14-12-17 V.Ivant (testem5-V10-04-00)
- corrected TestEm5.in to optimise output
	 
11-11-17 mma (testem5-V10-03-08)
- Updated README   .README.txt

07-11-17 V.Ivanchenko (testem5-V10-03-07)
- Updated fluo.mac
	 
28-07-17 SI (testem5-V10-03-06)
- Addition of dna.mac and update of pixe.mac

30-05-17 SI (testem5-V10-03-05)
- Addition of DNA constructors in PhysicsList
- Addition of DNA histograms in HistoManager
- Addition of flags in StackingAction
- Update of README

19-04-17 mma  (testem5-V10-03-04)
- BeginOfRunAction() : remove SetRandomNumberStore(true)

10-03-17 mma (testem5-V10-03-03)
- DetectorConstruction : correct definition of element Chrome

17-02-17 mma (testem5-V10-03-02)
- Run.cc : modify normalisation of histo 1 and 10

01-02-17 V.Ivanchenko (testem5-V10-03-01)
- PhysicsList - use more uniform builder names

27-12-16 S. Incerti (testem5-V10-03-00)
- updated pixe.mac

19-11-16 L.Urban (testem5-V10-02-07)
- typos corrected in Run.cc  

14-10-16 G.Folger (testem5-V10-02-06)
- remove direct use of {a,the}ParticleIterator, use GetParticleTableIterator().
    fix required by clang39 on Linux and MAC

28-07-16 mma  (testem5-V10-02-05)
- coding convention : rm double blank lines

04-07-16 I. Hrivnacova  (testem5-V10-02-04)
- Fixes for Doxygen documentation

11-04-16 mma  (testem5-V10-02-03)
- PhysListEmStandard: use G4EmParameters
- remove geant3 subdirectory

24-02-16 mma (testem5-V10-02-02)
- EventAction: correct filling of histo 1

11-01-16 V.Ivant (testem5-V10-02-01)
- PhysicsList - added hadron elastic physics

10-12-15 V.Ivant (testem5-V10-02-00)
- H.Burkhardt fixed root macro allowing to work both with root5 and 6
    
12-11-15 V.Ivant (testem5-V10-01-06)
- TrackingAction - more accurate fix which does not change results

11-11-15 V.Ivant (testem5-V10-01-05)
- TrackingAction - transmitted/reflected charged particles should
    have energy above zero to be included into histograms for
    transmitted/reflected particles energy and angles at exit

28-10-15 D.Sawkey (testem5-V10-01-04)
- update READMEs, again

27-10-15 D.Sawkey (testem5-V10-01-03)
- update physics list descriptions in .README, README

11-08-15 V.Ivant (testem5-V10-01-02)
- DetectorConstruction - do not remove volumes when sizes are changed
- SteppingVerbose - restore usage of this class
- TrackingAction - cleanup logic of transmitted/reflected particles

12-05-15 V.Ivant (testem5-V10-01-01)
- PhysicsList - use G4EmStandardPhysicsGS from physics_list library

04-03-15 V.Ivant (testem5-V10-01-00)
- Run - added forgotten fill of projectile scattering angle
    for printed output (printed value was zero)
	 
23-10-14 mma (testem5-V10-00-12)
- PhysListEmStandard : msc algorithm fUseSafetyPlus

21-10-14 mma (testem5-V10-00-11)
- PhysicsList : added G4EmStandardPhysicsSS, WVI

03-05-14 V.Ivant (testem5-V10-00-10)
- TestEm5.cc - added possibility to change number on threads by
    2nd parameter in command line, by default Michel's mechanism
    of usage of all cores is applied (previous tag); added
    printout of number of threads 
- StackingAction, HistoManager - added histograms for energy of
    gamma and Auger electrons per de-excitation model
	 
10-09-14 mma (testem5-V10-00-09)
- TestEm5.cc : SetNumberOfThreads( G4Threading::G4GetNumberOfCores() )

26-06-14 mma (testem5-V10-00-08)
- HistoManager - restaure "h" in front of histogram id
	 	 
16-06-14 mma (testem5-V10-00-07)
- remove PhysicsList::SetCuts()	 

07-06-14 mma (testem5-V10-00-06)
- minor cleanup in PhysicsList::SetCuts()
	 
04-06-14 mma (testem5-V10-00-05)
- rm plotHisto.kumac

03-05-14 V.Ivant (testem5-V10-00-04)
- TestEm5.cc - added possibility to change number on threads by
    2nd parameter in command line

03-05-14 V.Ivant (testem5-V10-00-03)
- Correct commit

02-05-14 V.Ivant (testem5-V10-00-02)

src/HistoManager.cc
//
// ********************************************************************
// * License and Disclaimer                                           *
// *                                                                  *
// * The  Geant4 software  is  copyright of the Copyright Holders  of *
// * the Geant4 Collaboration.  It is provided  under  the terms  and *
// * conditions of the Geant4 Software License,  included in the file *
// * LICENSE and available at  http://cern.ch/geant4/license .  These *
// * include a list of copyright holders.                             *
// *                                                                  *
// * Neither the authors of this software system, nor their employing *
// * institutes,nor the agencies providing financial support for this *
// * work  make  any representation or  warranty, express or implied, *
// * regarding  this  software system or assume any liability for its *
// * use.  Please see the license in the file  LICENSE  and URL above *
// * for the full disclaimer and the limitation of liability.         *
// *                                                                  *
// * This  code  implementation is the result of  the  scientific and *
// * technical work of the GEANT4 collaboration.                      *
// * By using,  copying,  modifying or  distributing the software (or *
// * any work based  on the software)  you  agree  to acknowledge its *
// * use  in  resulting  scientific  publications,  and indicate your *
// * acceptance of all terms of the Geant4 Software license.          *
// ********************************************************************
//
/// \file electromagnetic/TestEm5/src/HistoManager.cc
/// \brief Implementation of the HistoManager class
//
//
// 
//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......
//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

#include "HistoManager.hh"
#include "G4UnitsTable.hh"

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

HistoManager::HistoManager()
  : fFileName("sasidet")
{
  Book();
}

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

HistoManager::~HistoManager()
{
  delete G4AnalysisManager::Instance();
}

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

void HistoManager::Book()
{
  // Create or get analysis manager
  // The choice of analysis technology is done via selection of a namespace
  // in HistoManager.hh
  G4AnalysisManager* analysisManager = G4AnalysisManager::Instance();
  analysisManager->SetFileName(fFileName);
  analysisManager->SetVerboseLevel(1);
  analysisManager->SetActivation(true);    // enable inactivation of histograms

  // Define histograms start values
  const G4int kMaxHisto = 62;
  const G4String id[] = { "0", "1", "2", "3", "4", "5", "6", "7", "8", "9",
                         "10","11","12","13","14","15","16","17","18","19",
                         "20","21","22","23","24","25","26","27","28","29",
                         "30","31","32","33","34","35","36","37","38","39",
                         "40","41","42","43","44","45","46","47","48","49",
                         "50","51","52","53","54","55","56","57","58","59",
                         "60","61"
                        };
                        
  const G4String title[] =
                { "dummy",                                                //0
                  "energy deposit in absorber (not normalized)",          //1
                  "energy of charged secondaries at creation",            //2
                  "energy of neutral secondaries at creation",            //3
                  "energy of charged at creation (log scale)",            //4
                  "energy of neutral at creation (log scale)",            //5
                  "x_vertex of charged secondaries (all)",                //6
                  "x_vertex of charged secondaries (not absorbed)",       //7
                  "dummy","dummy",                                        //8-9
                  "(transmit, charged) : kinetic energy at exit: dN/dE",  //10
                  "(transmit, charged) : ener fluence: dE(MeV)/dOmega",   //11
                  "(transmit, charged) : space angle: dN/dOmega",         //12
                  "(transmit, charged) : projected angle at exit",        //13
                  "(transmit, charged) : projected position at exit",     //14
                  "(transmit, charged) : radius at exit",                 //15
                  "dummy",                                                //16
                  "dummy",                                                //17
                  "dummy",                                                //18
                  "dummy",                                                //19
                  "(transmit, neutral) : kinetic energy at exit",         //20
                  "(transmit, neutral) : ener fluence: dE(MeV)/dOmega",   //21
                  "(transmit, neutral) : space angle: dN/dOmega",         //22
                  "(transmit, neutral) : projected angle at exit",        //23
                  "dummy","dummy","dummy","dummy","dummy","dummy",       //24-29
                  "(reflect , charged) : kinetic energy at exit",         //30
                  "(reflect , charged) : ener fluence: dE(MeV)/dOmega",   //31
                  "(reflect , charged) : space angle: dN/dOmega",         //32
                  "(reflect , charged) : projected angle at exit",        //33
                  "dummy","dummy","dummy","dummy","dummy","dummy",       //34-39
                  "(reflect , neutral) : kinetic energy at exit",         //40
                  "(reflect , neutral) : ener fluence: dE(MeV)/dOmega",   //41
                  "(reflect , neutral) : space angle: dN/dOmega",         //42
                  "(reflect , neutral) : projected angle at exit",        //43
                  "dummy",                                                //44
                  "dummy",                                                //45
                  "dummy",                                                //46
                  "dummy",                                                //47
                  "dummy",                                                //48
                  "dummy",                                                //49
                  "energy of Auger e- at creation",                       //50
                  "energy of fluorescence gamma at creation",             //51
                  "energy of Auger e- at creation (log scale)",           //52
                  "energy of fluorescence gamma at creation (log scale)", //53
                  "energy of PIXE Auger e- at creation",                  //54
                  "energy of PIXE gamma at creation",                     //55
                  "energy of PIXE Auger e- at creation (log scale)",      //56
                  "energy of PIXE gamma at creation (log scale)",         //57
                  "energy of DNA Auger e- at creation",                   //58
                  "energy of DNA gamma at creation",                      //59
                  "energy of DNA Auger e- at creation (log scale)",       //60
                  "energy of DNA gamma at creation (log scale)"           //61
                 };

  // Default values (to be reset via /analysis/h1/set command)               
  G4int nbins = 100;
  G4double vmin = 0.;
  G4double vmax = 100.;

  // Create all histograms as inactivated 
  // as we have not yet set nbins, vmin, vmax
  for (G4int k=0; k<kMaxHisto; k++) {
    G4int ih = analysisManager->CreateH1("h"+id[k], title[k], nbins,vmin,vmax);
    analysisManager->SetH1Activation(ih, false);
  }
  
}
